ci-operator/step-registry/cucushift/installer/check/gcp/user-tags/cucushift-installer-check-gcp-user-tags-commands.sh

```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+if [ ! -f "${SHARED_DIR}/user_tags_sa.json" ]; then
+  echo "$(date -u --rfc-3339=seconds) - ERROR: Failed to find the key file of the IAM service-account for userTags testing on GCP."
+  exit 1
+fi
+
+if [ -f "${SHARED_DIR}/kubeconfig" ] ; then
+    export KUBECONFIG=${SHARED_DIR}/kubeconfig
+fi
+
+if [ -f "${SHARED_DIR}/proxy-conf.sh" ] ; then
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+INFRA_ID="$(oc get infrastructures.config.openshift.io cluster -o jsonpath='{.status.infrastructureName}')"
+GCP_REGION="${LEASED_RESOURCE}"
+
+GOOGLE_PROJECT_ID="$(< ${CLUSTER_PROFILE_DIR}/openshift_gcp_project)"
+export GCP_SHARED_CREDENTIALS_FILE="${SHARED_DIR}/user_tags_sa.json"
+sa_email=$(jq -r .client_email ${GCP_SHARED_CREDENTIALS_FILE})
+if ! gcloud auth list | grep -E "\*\s+${sa_email}"
+then
+  gcloud auth activate-service-account --key-file="${GCP_SHARED_CREDENTIALS_FILE}"
+  gcloud config set project "${GOOGLE_PROJECT_ID}"
+fi
+
+# User-defined tags validation. It will check if each user-defined tag is applied. 
+# Return non-zero is one or more user-defined tag absent. 
+# $1 - the current tags of the resource under question
+function validate_user_tags() {
+  local -r current_tags="$1";  shift
+
+  printf '%s' "${USER_TAGS:-}" | while read -r PARENT KEY VALUE || [ -n "${PARENT}" ]
+  do
+    a_tag_value="${PARENT}/${KEY}/${VALUE}"
+    space_pattern="\ "
+    if [[ "${a_tag_value}" =~ ${space_pattern} ]] && [ ${#a_tag_value} -gt 64 ]; then
+      echo "$(date -u --rfc-3339=seconds) - Truncating tag '${a_tag_value}'"
+      a_tag_value="${a_tag_value::63}"
+      a_tag_value="${a_tag_value%* }"
```

Comment:
I know the above line is to get a truncated string including the leading 63 chars, what is the purpose of this line?

Reply:
To remove the any trailing space.

Reply:
Seem like that only trim one tailing space.
```
$ a_tag_value="aaa bbb   "
$ echo "${a_tag_value% *}==="
aaa bbb  ===
```
How about using `a_tag_value=$(echo "${a_tag_value}" | xargs)`

Reply:
@jianlinliu I just tried a bit, and it looks like no word wrapping for multiple spaces, i.e. multiple continuous spaces would be taken a word. That said, truncating only the ending space, in case the space is rightly after a word, should be ok. Your suggestion? Thanks!  
```
---
namespacedTagKey: openshift-qe/Su.Shi-Die_Lian_Hua-Chun_Jing
namespacedTagValue: openshift-qe/Su.Shi-Die_Lian_Hua-Chun_Jing/aaa   bbb   ccc   ddd   eee   fff   ggg   hhh   iii   kkk
tagKey: tagKeys/281475274599629
tagKeyParentName: projects/1042363005003
tagValue: tagValues/281484366183786

Reply:
If we use multiple spaces instead of one single space for the tag value in the provision chain, we even do not take text wrapping into consideration, right?

Reply:
If you are talking about the behavior of the gcloud command output, yes, no word wrapping for multiple spaces.

Reply:
as a comprise, how about directly setting https://github.com/openshift/release/pull/47844/files#diff-9e16ef16dd005adea4132e04cfbc9c7caa4b36193b1e969d21e57439c75c7989R22 with multiple spaces, then we do not need any code for text wrapping case.

Reply:
Do you mean to remove all values which lead to truncation? For example, 
```
---
namespacedTagKey: XXXXXXXXXXXX/2023.begin-with-numerical_character
namespacedTagValue: XXXXXXXXXXXX/2023.begin-with-numerical_character/2023 begin-with
  numerical_character
tagKey: tagKeys/281483130531336
tagKeyParentName: projects/1042363005003
tagValue: tagValues/281483456805293
---
namespacedTagKey: XXXXXXXXXXXX/numerical_character.2023-in-middle
namespacedTagValue: XXXXXXXXXXXX/numerical_character.2023-in-middle/numerical_character
  2023-in-middle
tagKey: tagKeys/281480468513933
tagKeyParentName: projects/1042363005003
tagValue: tagValues/281477984235945
---
namespacedTagKey: XXXXXXXXXXXX/Li.Bai
namespacedTagValue: XXXXXXXXXXXX/Li.Bai/chuang qian ming yue guang, yi shi di shang
  shuang
tagKey: tagKeys/281482700904696
tagKeyParentName: projects/1042363005003
tagValue: tagValues/281482854611452
---
namespacedTagKey: XXXXXXXXXXXX/Su.Shi-Die_Lian_Hua-Chun_Jing
namespacedTagValue: XXXXXXXXXXXX/Su.Shi-Die_Lian_Hua-Chun_Jing/Hua Tui Can Hong Qing
  Xing Xiao
tagKey: tagKeys/281475274599629
tagKeyParentName: projects/1042363005003
tagValue: tagValues/281477497267603
---
namespacedTagKey: XXXXXXXXXXXX/uppercase_character-IN-middle
namespacedTagValue: XXXXXXXXXXXX/uppercase_character-IN-middle/uppercase_CHARACTER
  IN-middle
tagKey: tagKeys/281480976057898
tagKeyParentName: projects/1042363005003
tagValue: tagValues/281475523161733
---
namespacedTagKey: XXXXXXXXXXXX/end-with.numerical_character-2023
namespacedTagValue: XXXXXXXXXXXX/end-with.numerical_character-2023/end-with numerical_character
  2023
tagKey: tagKeys/281480875755836
tagKeyParentName: projects/1042363005003
tagValue: tagValues/281479420147043
---
namespacedTagKey: XXXXXXXXXXXX/Su.Shi-Ti_Xi_Lin_Bi
namespacedTagValue: XXXXXXXXXXXX/Su.Shi-Ti_Xi_Lin_Bi/Heng Kan Cheng Ling Ce Cheng
  Feng
tagKey: tagKeys/281481549282667
tagKeyParentName: projects/1042363005003
tagValue: tagValues/281477522401639
```

```
2024-01-22 03:23:35+00:00 - Truncating tag 'XXXXXXXXXXXX/Su.Shi-Die_Lian_Hua-Chun_Jing/Hua Tui Can Hong Qing Xing Xiao'
2024-01-22 03:23:35+00:00 - Truncating tag 'XXXXXXXXXXXX/numerical_character.2023-in-middle/numerical_character 2023-in-middle'
2024-01-22 03:23:35+00:00 - Truncating tag 'XXXXXXXXXXXX/end-with.numerical_character-2023/end-with numerical_character 2023'
2024-01-22 03:23:35+00:00 - Truncating tag 'XXXXXXXXXXXX/uppercase_character-IN-middle/uppercase_CHARACTER IN-middle'
2024-01-22 03:23:35+00:00 - Truncating tag 'XXXXXXXXXXXX/Su.Shi-Ti_Xi_Lin_Bi/Heng Kan Cheng Ling Ce Cheng Feng'
2024-01-22 03:23:35+00:00 - Truncating tag 'XXXXXXXXXXXX/Li.Bai/chuang qian ming yue guang, yi shi di shang shuang'
2024-01-22 03:23:35+00:00 - Truncating tag 'XXXXXXXXXXXX/2023.begin-with-numerical_character/2023 begin-with numerical_character'

Reply:
okay, I did not notices so many tag are being Truncated, then I will let it go.